Race Condition in job scheduler

Race Condition


```go
package internal

import (
	"fmt"
	"sync"
	"testing"
	"time"

	"github.com/go-co-op/gocron"
)

var (
	cronMutex sync.Mutex
)

// AddCronJob enqueues a cron job into the scheduler.
func AddCronJob(scheduler *gocron.Scheduler, f func() error) error {
	jobSpec := scheduler.Cron("0 0 1 * *").SingletonMode().Tag("test").StartImmediately()
	_, jobErr := jobSpec.Do(func() error {
		cronMutex.Lock()
		defer cronMutex.Unlock()

		if err := f(); err != nil {
			return err
		}
		return nil
	})

	return jobErr
}

func TestAddCronJob(t *testing.T) {
	scheduler := gocron.NewScheduler(time.UTC)
	scheduler.SetMaxConcurrentJobs(1, gocron.RescheduleMode)
	defer scheduler.Stop()

	if err := AddCronJob(scheduler, func() error {
		return fmt.Errorf("Angry job")
	}); err != nil {
		t.Fatal(err)
	}

	i := 0
	ch := make(chan bool, 1)
	if err := AddCronJob(scheduler, func() error {
		i++
		ch <- true
		return nil
	}); err != nil {
		t.Fatal(err)
	}

	scheduler.StartAsync()
	<-ch
	scheduler.Stop()

	if i != 1 {
		t.Errorf("counter mismatch, got: %d want: %d", i, 1)
	}
	close(ch)
}
```

```bash
CGO_ENABLED=1 go test -race internal -count 10

==================
WARNING: DATA RACE
Write at 0x00c00012a448 by goroutine 34:
  github.com/go-co-op/gocron.(*jobFunction).singletonRunner()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.1/executor.go:89 +0x2bc
  github.com/go-co-op/gocron.(*executor).runJob.func2()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.1/executor.go:167 +0x39

Previous read at 0x00c00012a448 by goroutine 30:
  github.com/go-co-op/gocron.(*executor).runJob()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.1/executor.go:169 +0x230
  github.com/go-co-op/gocron.(*executor).limitModeRunner()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.1/executor.go:109 +0x1e4
  github.com/go-co-op/gocron.(*executor).run.func2()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.1/executor.go:187 +0x39

Goroutine 34 (running) created at:
  github.com/go-co-op/gocron.(*executor).runJob()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.1/executor.go:167 +0x219
  github.com/go-co-op/gocron.(*executor).limitModeRunner()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.1/executor.go:109 +0x1e4
  github.com/go-co-op/gocron.(*executor).run.func2()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.1/executor.go:187 +0x39

Goroutine 30 (finished) created at:
  github.com/go-co-op/gocron.(*executor).run()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.1/executor.go:187 +0x3be
  github.com/go-co-op/gocron.(*executor).start.func1()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.1/executor.go:130 +0x39
==================
--- FAIL: TestAddCronJob (0.00s)
    testing.go:1446: race detected during execution of test
```

That one got fixed but I have another error on stop now. I'm running on a Linux VM using Go 1.20.4.

```bash
CGO_ENABLED=1 go test -race internal -count 100
==================
WARNING: DATA RACE
Write at 0x00c00001a2b8 by goroutine 306:
  runtime.racewrite()
      <autogenerated>:1 +0x24
  github.com/go-co-op/gocron.(*executor).stop.func1()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.3-0.20230602132527-7390192a00bf/executor.go:247 +0x47
  sync.(*Map).Range()
      /usr/local/go/src/sync/map.go:476 +0x1db
  github.com/go-co-op/gocron.(*executor).stop()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.3-0.20230602132527-7390192a00bf/executor.go:245 +0xd3
  github.com/go-co-op/gocron.(*Scheduler).stop()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.3-0.20230602132527-7390192a00bf/scheduler.go:900 +0xb2
  github.com/go-co-op/gocron.(*Scheduler).Stop()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.3-0.20230602132527-7390192a00bf/scheduler.go:894 +0x56
  internal.TestAddCronJob()
      /internal/cronfail_test.go:69 +0x3a4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /usr/local/go/src/testing/testing.go:1629 +0x47

Previous read at 0x00c00001a2b8 by goroutine 312:
  runtime.raceread()
      <autogenerated>:1 +0x24
  github.com/go-co-op/gocron.(*jobFunction).singletonRunner()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.3-0.20230602132527-7390192a00bf/executor.go:83 +0xb1
  github.com/go-co-op/gocron.(*executor).runJob.func2()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.3-0.20230602132527-7390192a00bf/executor.go:169 +0x39

Goroutine 306 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1629 +0x805
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:2036 +0x8d
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1576 +0x216
  testing.runTests()
      /usr/local/go/src/testing/testing.go:2034 +0x87c
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1906 +0xb44
  main.main()
      _testmain.go:47 +0x2e9

Goroutine 312 (finished) created at:
  github.com/go-co-op/gocron.(*executor).runJob()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.3-0.20230602132527-7390192a00bf/executor.go:169 +0x21c
  github.com/go-co-op/gocron.(*executor).limitModeRunner()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.3-0.20230602132527-7390192a00bf/executor.go:111 +0x206
  github.com/go-co-op/gocron.(*executor).run.func2()
      /home/coder/go/pkg/mod/github.com/go-co-op/gocron@v1.28.3-0.20230602132527-7390192a00bf/executor.go:191 +0x39
==================
--- FAIL: TestAddCronJob (0.00s)
    testing.go:1446: race detected during execution of test
FAIL
FAIL    internal     0.039s
FAIL
```
```bash
go version
go version go1.20.4 linux/amd64

uname -a
Linux a5c36d205060 5.15.0-72-generic #79-Ubuntu SMP Wed Apr 19 08:22:18 UTC 2023 x86_64 GNU/Linux

lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 11 (bullseye)
Release:        11
Codename:       bullseye
```

--- a/executor.go
+++ b/executor.go
@@ -80,7 +80,9 @@ func runJob(f jobFunction) {
 
 func (jf *jobFunction) singletonRunner() {
 	jf.singletonRunnerOn.Store(true)
+	jf.singletonWgMu.Lock()
 	jf.singletonWg.Add(1)
+	jf.singletonWgMu.Unlock()
 	for {
 		select {
 		case <-jf.ctx.Done():
@@ -163,7 +165,7 @@ func (e *executor) runJob(f jobFunction) {
 		}
 		runJob(f)
 	case singletonMode:
-		e.singletonWgs.Store(f.singletonWg, struct{}{})
+		e.singletonWgs.Store(f.singletonWg, f.singletonWgMu)
 
 		if !f.singletonRunnerOn.Load() {
 			go f.singletonRunner()
@@ -243,8 +245,12 @@ func (e *executor) stop() {
 	e.wg.Wait()
 	if e.singletonWgs != nil {
 		e.singletonWgs.Range(func(key, value interface{}) bool {
-			if wg, ok := key.(*sync.WaitGroup); ok {
+			wg, wgOk := key.(*sync.WaitGroup)
+			mu, muOk := value.(*sync.Mutex)
+			if wgOk && muOk {
+				mu.Lock()
 				wg.Wait()
+				mu.Unlock()
 			}
 			return true
 		})

--- a/job.go
+++ b/job.go
@@ -58,6 +58,7 @@ type jobFunction struct {
 	runStartCount     *atomic.Int64      // number of times the job was started
 	runFinishCount    *atomic.Int64      // number of times the job was finished
 	singletonWg       *sync.WaitGroup    // used by singleton runner
+	singletonWgMu     *sync.Mutex        // use to protect the singletonWg
 	stopped           *atomic.Bool       // tracks whether the job is currently stopped
 	jobFuncNextRun    time.Time          // the next time the job is scheduled to run
 }
@@ -88,6 +89,7 @@ func (jf *jobFunction) copy() jobFunction {
 		runStartCount:     jf.runStartCount,
 		runFinishCount:    jf.runFinishCount,
 		singletonWg:       jf.singletonWg,
+		singletonWgMu:     jf.singletonWgMu,
 		singletonRunnerOn: jf.singletonRunnerOn,
 		stopped:           jf.stopped,
 		jobFuncNextRun:    jf.jobFuncNextRun,
@@ -422,11 +424,16 @@ func (j *Job) SingletonMode() {
 	j.mu.Lock()
 	defer j.mu.Unlock()
 	j.runConfig.mode = singletonMode
+
+	j.jobFunction.singletonWgMu = &sync.Mutex{}
+	j.jobFunction.singletonWgMu.Lock()
 	j.jobFunction.singletonWg = &sync.WaitGroup{}
-	j.singletonQueueMu = &sync.Mutex{}
-	j.singletonQueueMu.Lock()
-	defer j.singletonQueueMu.Unlock()
+	j.jobFunction.singletonWgMu.Unlock()
+
+	j.jobFunction.singletonQueueMu = &sync.Mutex{}
+	j.jobFunction.singletonQueueMu.Lock()
 	j.jobFunction.singletonQueue = make(chan struct{}, 100)
+	j.jobFunction.singletonQueueMu.Unlock()
 }
 
 // shouldRun evaluates if this job should run again

